File: scripts/doc_coverage_report.js
#!/usr/bin/env node
/**
 * Cross-reference March 4 planner log against documentation coverage.
 * Generates data/daily_doc_coverage_2026-03-04.txt
 */

const fs = require('fs');
const path = require('path');

const BASE = path.join('C:', 'Users', 'kurt_', '.openclaw', 'workspace', 'kontensystem');

// ── Load data ─────────────────────────────────────────────────────────────────

const logData = JSON.parse(fs.readFileSync(path.join(BASE, 'logs', 'planner-log-2026-03-04.json'), 'utf8'));

const taskByCode = {};   // code → task entry
const taskByName = {};   // lower-stripped name → task entry

const mtlLines = fs.readFileSync(path.join(BASE, 'data', 'master_task_list_v4.jsonl'), 'utf8').split('\n');
for (const line of mtlLines) {
  const l = line.trim();
  if (!l) continue;
  const t = JSON.parse(l);
  if (t.code) taskByCode[t.code] = t;
  const key = t.name.toLowerCase().trim();
  if (!taskByName[key]) taskByName[key] = t;
}

const docIndex = {};  // doc_id → doc entry
const diLines = fs.readFileSync(path.join(BASE, 'data', 'documentation_index.jsonl'), 'utf8').split('\n');
for (const line of diLines) {
  const l = line.trim();
  if (!l) continue;
  const d = JSON.parse(l);
  docIndex[d.doc_id] = d;
}

const TODAY = new Date('2026-03-05');

function docAgeYears(doc) {
  const lm = doc.last_modified || '';
  if (!lm) return null;
  // Try YYYY-MM-DD prefix
  const m = lm.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (m) {
    const d = new Date(`${m[1]}-${m[2]}-${m[3]}`);
    if (!isNaN(d)) {
      return Math.round(10 * (TODAY - d) / (365.25 * 24 * 3600 * 1000)) / 10;
    }
  }
  return null;
}

// ── Extract 6-char code from activity string ──────────────────────────────────

function extractCode(activity) {
  const m = activity.match(/\b([A-Z]{2,6}[A-Z0-9]{0,4})\s*(?:\([^)]*\))?\s*$/);
  if (m && m[1].length === 6) return m[1];
  return '';
}

// ── Match log entry to master task ────────────────────────────────────────────

function matchTask(entry) {
  const code = extractCode(entry.activity);
  if (code && taskByCode[code]) return [taskByCode[code], 'code'];
  const orig = entry.original_activity || '';
  if (orig) {
    const code2 = extractCode(orig);
    if (code2 && taskByCode[code2]) return [taskByCode[code2], 'orig_code'];
  }
  const nameClean = entry.activity.replace(/\s+[A-Z]{6}\s*(?:\([^)]*\))?\s*$/, '').trim().toLowerCase();
  if (taskByName[nameClean]) return [taskByName[nameClean], 'name_exact'];
  if (nameClean.length >= 10) {
    const prefix20 = nameClean.slice(0, 20);
    for (const [key, t] of Object.entries(taskByName)) {
      if (key.slice(0, 20) === prefix20) return [t, 'name_prefix'];
    }
  }
  return [null, null];
}

// ── Compute actual minutes ─────────────────────────────────────────────────────

function actualMinutes(entry) {
  try {
    const [sh, sm, ss] = entry.started_at.split(':').map(Number);
    const [ch, cm, cs] = entry.completed_at.split(':').map(Number);
    const startSec = sh * 3600 + sm * 60 + ss;
    const endSec = ch * 3600 + cm * 60 + cs;
    let diff = Math.floor((endSec - startSec) / 60);
    if (diff < 0) diff += 24 * 60;
    return diff;
  } catch {
    return entry.minutes || 0;
  }
}

// ── Process log entries ────────────────────────────────────────────────────────

const records = [];
for (const entry of logData) {
  if (entry.skipped) continue;
  const mins = actualMinutes(entry);
  const [task, method] = matchTask(entry);
  const code = extractCode(entry.activity) || extractCode(entry.original_activity || '') || '';

  let docStatus, docRefs, accountPrefix;
  if (task) {
    docStatus = task.doc_status || 'undocumented';
    docRefs = task.doc_refs || [];
    accountPrefix = task.account_prefix || '';
  } else {
    docStatus = 'unmatched';
    docRefs = [];
    accountPrefix = '';
  }

  const docDetails = [];
  for (const ref of docRefs) {
    const doc = docIndex[ref.doc_id];
    if (doc) {
      docDetails.push({
        doc_id: ref.doc_id,
        title: doc.title || ref.doc_id,
        last_modified: doc.last_modified || '?',
        age_years: docAgeYears(doc),
        match_type: ref.match_type || '',
      });
    }
  }

  records.push({
    activity: entry.activity,
    code,
    list: entry.list || '',
    minutes: mins,
    started_at: entry.started_at,
    completed_at: entry.completed_at,
    doc_status: docStatus,
    match_method: method,
    account_prefix: accountPrefix,
    doc_details: docDetails,
    task_name: task ? task.name : '',
  });
}

// ── Aggregate by canonical key (code or stripped name) ─────────────────────────

function canonicalKey(rec) {
  if (rec.code) return rec.code;
  return rec.activity.replace(/\s+[A-Z]{6}\s*(?:\([^)]*\))?\s*$/, '').trim();
}

const STATUS_RANK = { documented: 3, mentioned: 2, undocumented: 1, unmatched: 0 };

const aggregated = {};
for (const rec of records) {
  const key = canonicalKey(rec);
  if (!aggregated[key]) {
    aggregated[key] = {
      activities: [], minutes: 0, doc_status: 'unmatched',
      code: '', account_prefix: '', doc_details: [], task_name: '',
    };
  }
  const agg = aggregated[key];
  agg.activities.push(rec.activity);
  agg.minutes += rec.minutes;
  if ((STATUS_RANK[rec.doc_status] || 0) > (STATUS_RANK[agg.doc_status] || 0)) {
    agg.doc_status = rec.doc_status;
    agg.account_prefix = rec.account_prefix;
    agg.task_name = rec.task_name;
  }
  if (!agg.code && rec.code) agg.code = rec.code;
  const existingIds = new Set(agg.doc_details.map(d => d.doc_id));
  for (const d of rec.doc_details) {
    if (!existingIds.has(d.doc_id)) {
      agg.doc_details.push(d);
      existingIds.add(d.doc_id);
    }
  }
}

// Convert to array, sort by minutes desc
let items = Object.entries(aggregated).map(([key, agg]) => {
  const activity = agg.activities[0];
  let display = activity.replace(/\s+[A-Z]{6}\s*(?:\([^)]*\))?\s*$/, '').trim();
  if (!display) display = activity;
  return {
    key,
    display,
    code: agg.code,
    minutes: agg.minutes,
    doc_status: agg.doc_status,
    account_prefix: agg.account_prefix,
    task_name: agg.task_name,
    doc_details: agg.doc_details,
    occurrences: agg.activities.length,
  };
});

items.sort((a, b) => b.minutes - a.minutes);

// ── Statistics ─────────────────────────────────────────────────────────────────

const totalMins = records.reduce((s, r) => s + r.minutes, 0);
const byStatus = {};
for (const r of records) {
  byStatus[r.doc_status] = (byStatus[r.doc_status] || 0) + r.minutes;
}

const documentedMins = byStatus['documented'] || 0;
const mentionedMins = byStatus['mentioned'] || 0;
const undocumentedMins = byStatus['undocumented'] || 0;
const unmatchedMins = byStatus['unmatched'] || 0;

const pct = n => totalMins ? (100 * n / totalMins).toFixed(1) : '0.0';

// ── Oldest docs ────────────────────────────────────────────────────────────────

const seenDocIds = {};
for (const item of items) {
  for (const dd of item.doc_details) {
    if (dd.age_years !== null && !seenDocIds[dd.doc_id]) {
      seenDocIds[dd.doc_id] = { ...dd, activity: item.display, activity_mins: item.minutes };
    }
  }
}
const oldestDocs = Object.values(seenDocIds)
  .sort((a, b) => b.age_years - a.age_years)
  .slice(0, 10);

// Biggest undoc time sinks
const undocItems = items.filter(i => ['undocumented', 'unmatched'].includes(i.doc_status) && i.minutes > 0);
undocItems.sort((a, b) => b.minutes - a.minutes);

// ── Format helpers ─────────────────────────────────────────────────────────────

const STATUS_ICON = { documented: '✅', mentioned: '⚠️ ', undocumented: '❌', unmatched: '🔍' };
const STATUS_LABEL = { documented: 'DOCUMENTED', mentioned: 'MENTIONED ', undocumented: 'UNDOC     ', unmatched: 'UNMATCHED ' };

function pad(s, n, right = false) {
  s = String(s);
  // For strings with emoji/unicode, rough length
  const vis = s.replace(/[^\x00-\x7F]/g, '  '); // treat non-ASCII as 2 chars
  const len = vis.length;
  if (len >= n) return s;
  const pad = ' '.repeat(n - len);
  return right ? pad + s : s + pad;
}

function truncate(s, n) {
  if (s.length > n) return s.slice(0, n - 1) + '…';
  return s;
}

const out = [];

out.push('='.repeat(80));
out.push('DAILY DOCUMENTATION COVERAGE REPORT — March 4, 2026');
out.push('='.repeat(80));
out.push('Generated: 2026-03-05');
out.push('');

out.push('─'.repeat(80));
out.push('SUMMARY STATISTICS');
out.push('─'.repeat(80));
out.push(`  Total minutes logged        : ${String(totalMins).padStart(4)} min`);
out.push(`  Documented (✅)            : ${String(documentedMins).padStart(4)} min  (${pct(documentedMins).padStart(5)}%)`);
out.push(`  Mentioned  (⚠️)             : ${String(mentionedMins).padStart(4)} min  (${pct(mentionedMins).padStart(5)}%)`);
out.push(`  Undocumented (❌)          : ${String(undocumentedMins).padStart(4)} min  (${pct(undocumentedMins).padStart(5)}%)`);
out.push(`  Unmatched (🔍)             : ${String(unmatchedMins).padStart(4)} min  (${pct(unmatchedMins).padStart(5)}%)`);
out.push(`  Coverage (doc + mentioned)  :        ${pct(documentedMins + mentionedMins).padStart(5)}%`);
out.push('');

out.push('─'.repeat(80));
out.push('FULL ACTIVITY LIST  (sorted by time spent, most first)');
out.push('─'.repeat(80));

for (const item of items) {
  const icon = STATUS_ICON[item.doc_status] || '?';
  const label = STATUS_LABEL[item.doc_status] || item.doc_status;
  const display = truncate(item.display, 44).padEnd(44);
  const codeStr = (item.code || '      ').padEnd(6);
  const minsStr = String(item.minutes).padStart(4);

  if (item.doc_details.length > 0) {
    // Prefer named matches
    const named = item.doc_details.filter(d => d.match_type === 'named');
    let best = named.length > 0 ? named : item.doc_details;
    // Deduplicate by title
    const seenTitles = new Set();
    const dedupedBest = [];
    for (const d of best) {
      if (!seenTitles.has(d.title)) { seenTitles.add(d.title); dedupedBest.push(d); }
    }
    best = dedupedBest.slice(0, 3);

    let first = true;
    for (const d of best) {
      const ageStr = d.age_years !== null ? `${d.age_years.toFixed(1)}y` : '?y';
      const titleShort = truncate(d.title, 38);
      if (first) {
        out.push(`  ${display} ${codeStr}  ${minsStr}  ${icon} ${label}  [${d.doc_id}] ${titleShort} (${ageStr})`);
        first = false;
      } else {
        out.push(`  ${''.padEnd(44)} ${''.padEnd(6)}  ${''.padStart(4)}  ${''.padEnd(13)}  [${d.doc_id}] ${titleShort} (${ageStr})`);
      }
    }
  } else {
    let note = '';
    if (item.doc_status === 'unmatched') {
      note = item.code ? `→ code ${item.code} not in task list` : '→ not in task list';
    } else if (item.doc_status === 'undocumented') {
      note = item.account_prefix ? `→ acct: ${item.account_prefix}` : '→ no account';
    } else if (item.doc_status === 'mentioned') {
      note = `→ acct: ${item.account_prefix}`;
    }
    out.push(`  ${display} ${codeStr}  ${minsStr}  ${icon} ${label}  ${note}`);
  }
}
out.push('');

out.push('─'.repeat(80));
out.push('BIGGEST UNDOCUMENTED / UNMATCHED TIME SINKS  (top 15)');
out.push('─'.repeat(80));
for (let i = 0; i < Math.min(15, undocItems.length); i++) {
  const item = undocItems[i];
  const display = truncate(item.display, 50).padEnd(50);
  const codeStr = item.code ? `[${item.code}]` : '[------]';
  const status = item.doc_status.toUpperCase();
  const acct = item.account_prefix ? `acct=${item.account_prefix}` : 'no acct';
  out.push(`  ${String(i + 1).padStart(2)}. ${display} ${codeStr}  ${String(item.minutes).padStart(4)} min  ${status}  ${acct}`);
}
out.push('');

out.push('─'.repeat(80));
out.push('OLDEST DOCUMENTATION REFERENCED TODAY  (top 10 by age)');
out.push('─'.repeat(80));
if (oldestDocs.length > 0) {
  for (const d of oldestDocs) {
    const titleShort = truncate(d.title, 55);
    out.push(`  [${d.doc_id}] ${titleShort}`);
    out.push(`       Last modified: ${d.last_modified}  (age: ${d.age_years.toFixed(1)} years)  [${d.match_type}]`);
    out.push(`       Referenced by: ${d.activity} (${d.activity_mins} min)`);
    out.push('');
  }
} else {
  out.push('  (No dated documentation found)');
  out.push('');
}

out.push('─'.repeat(80));
out.push('NOTES');
out.push('─'.repeat(80));
out.push('  Status definitions:');
out.push('    ✅ DOCUMENTED  — task has a named or direct doc reference in master task list');
out.push('    ⚠️  MENTIONED   — task is in an account-level doc but not specifically described');
out.push('    ❌ UNDOCUMENTED — task has no doc references at all');
out.push('    🔍 UNMATCHED   — activity could not be matched to any task in master_task_list_v4');
out.push('');
out.push('  Mentioned docs are typically account-level documents that cover the area');
out.push('  broadly (e.g. Dokumentation Lebenserhaltung) but don\'t describe this specific');
out.push('  task. They count as weak coverage.');
out.push('');

const reportPath = path.join(BASE, 'data', 'daily_doc_coverage_2026-03-04.txt');
fs.writeFileSync(reportPath, out.join('\n') + '\n', 'utf8');

console.log(`Report written: ${reportPath}`);
console.log();
console.log('=== QUICK SUMMARY ===');
console.log(`Total minutes logged : ${totalMins}`);
console.log(`✅ Documented        : ${documentedMins} min (${pct(documentedMins)}%)`);
console.log(`⚠️  Mentioned         : ${mentionedMins} min (${pct(mentionedMins)}%)`);
console.log(`❌ Undocumented      : ${undocumentedMins} min (${pct(undocumentedMins)}%)`);
console.log(`🔍 Unmatched         : ${unmatchedMins} min (${pct(unmatchedMins)}%)`);
console.log(`Coverage (doc+ment.) : ${pct(documentedMins + mentionedMins)}%`);
console.log();
console.log('Top 5 undocumented/unmatched time sinks:');
for (const item of undocItems.slice(0, 5)) {
  console.log(`  ${String(item.minutes).padStart(4)} min  [${(item.code || '------').padEnd(6)}]  ${item.display.slice(0, 55)}`);
}

File: scripts/doc_coverage_report.py
#!/usr/bin/env python3
"""
Cross-reference March 4 planner log against documentation coverage.
Generates data/daily_doc_coverage_2026-03-04.txt
"""
import json
import re
import os
from datetime import datetime, date
from collections import defaultdict

BASE = r"C:\Users\kurt_\.openclaw\workspace\kontensystem"

# ── Load data ─────────────────────────────────────────────────────────────────

with open(os.path.join(BASE, "logs", "planner-log-2026-03-04.json"), encoding="utf-8") as f:
    log = json.load(f)

task_by_code = {}   # code → task entry
task_by_name = {}   # lower-stripped name → task entry

with open(os.path.join(BASE, "data", "master_task_list_v4.jsonl"), encoding="utf-8") as f:
    for line in f:
        line = line.strip()
        if not line:
            continue
        t = json.loads(line)
        if t.get("code"):
            task_by_code[t["code"]] = t
        # also index by first 20 chars of name (lowered) for fuzzy name matching
        key = t["name"].lower().strip()
        if key not in task_by_name:
            task_by_name[key] = t

doc_index = {}  # doc_id → doc entry
with open(os.path.join(BASE, "data", "documentation_index.jsonl"), encoding="utf-8") as f:
    for line in f:
        line = line.strip()
        if not line:
            continue
        d = json.loads(line)
        doc_index[d["doc_id"]] = d

TODAY = date(2026, 3, 5)   # report generation date

def doc_age_years(doc):
    """Return age in years from last_modified string (YYYY-MM-DD or similar)."""
    lm = doc.get("last_modified", "")
    if not lm:
        return None
    # Try YYYY-MM-DD
    for fmt in ("%Y-%m-%d", "%Y-%m-%dT%H:%M:%S", "%d.%m.%Y"):
        try:
            d = datetime.strptime(lm[:10], fmt[:len(lm[:10])]).date()
            return round((TODAY - d).days / 365.25, 1)
        except Exception:
            pass
    # Try numeric year embedded in title like "7.5.29" → 2007 or 2029?
    # last_modified field should be a proper date; if unparseable, return None
    return None

# ── Extract 6-char code from activity string ──────────────────────────────────

def extract_code(activity: str):
    """Return the last 6-char uppercase alphanum token if it looks like a code."""
    m = re.search(r'\b([A-Z]{2,6}[A-Z0-9]{0,4})\s*(?:\([^)]*\))?\s*$', activity)
    if m:
        cand = m.group(1)
        if len(cand) == 6:
            return cand
    return None

# ── Match log entry to master task ────────────────────────────────────────────

def match_task(entry):
    """Return (task_entry_or_None, match_method)"""
    activity = entry["activity"]
    code = extract_code(activity)
    if code and code in task_by_code:
        return task_by_code[code], "code"
    # Try original_activity code
    orig = entry.get("original_activity", "")
    if orig:
        code2 = extract_code(orig)
        if code2 and code2 in task_by_code:
            return task_by_code[code2], "orig_code"
    # Try name match (strip code suffix, lowercase)
    name_clean = re.sub(r'\s+[A-Z]{6}\s*(\([^)]*\))?\s*$', '', activity).strip().lower()
    if name_clean in task_by_name:
        return task_by_name[name_clean], "name_exact"
    # Try prefix match (first 30 chars)
    for key, t in task_by_name.items():
        if len(name_clean) >= 10 and name_clean[:20] == key[:20]:
            return t, "name_prefix"
    return None, None

# ── Process log entries ────────────────────────────────────────────────────────

def actual_minutes(entry):
    """Compute actual minutes from started_at / completed_at."""
    try:
        s = datetime.strptime(entry["started_at"], "%H:%M:%S")
        c = datetime.strptime(entry["completed_at"], "%H:%M:%S")
        diff = (c - s).seconds // 60
        if diff < 0:
            diff += 24 * 60
        return diff
    except Exception:
        return entry.get("minutes", 0)

records = []
for entry in log:
    if entry.get("skipped"):
        continue
    mins = actual_minutes(entry)
    task, method = match_task(entry)
    code = extract_code(entry["activity"]) or (extract_code(entry.get("original_activity", "")) or "")
    
    if task:
        doc_status = task.get("doc_status", "undocumented")
        doc_refs = task.get("doc_refs", [])
        account_prefix = task.get("account_prefix", "")
    else:
        doc_status = "unmatched"
        doc_refs = []
        account_prefix = ""

    # Build doc details
    doc_details = []
    for ref in doc_refs:
        did = ref["doc_id"]
        doc = doc_index.get(did)
        if doc:
            age = doc_age_years(doc)
            doc_details.append({
                "doc_id": did,
                "title": doc.get("title", did),
                "last_modified": doc.get("last_modified", "?"),
                "age_years": age,
                "match_type": ref.get("match_type", ""),
            })

    records.append({
        "activity": entry["activity"],
        "code": code,
        "list": entry.get("list", ""),
        "minutes": mins,
        "started_at": entry["started_at"],
        "completed_at": entry["completed_at"],
        "doc_status": doc_status,
        "match_method": method,
        "account_prefix": account_prefix,
        "doc_details": doc_details,
        "task_name": task["name"] if task else "",
    })

# ── Aggregate: group by canonical activity (code or stripped name) ─────────────

def canonical_key(rec):
    if rec["code"]:
        return rec["code"]
    return re.sub(r'\s+[A-Z]{6}\s*(\([^)]*\))?\s*$', '', rec["activity"]).strip()

aggregated = defaultdict(lambda: {
    "activities": [], "minutes": 0, "doc_status": "unmatched",
    "code": "", "account_prefix": "", "doc_details": [], "task_name": ""
})

for rec in records:
    key = canonical_key(rec)
    agg = aggregated[key]
    agg["activities"].append(rec["activity"])
    agg["minutes"] += rec["minutes"]
    # Keep the most informative doc_status
    status_rank = {"documented": 3, "mentioned": 2, "undocumented": 1, "unmatched": 0}
    if status_rank.get(rec["doc_status"], 0) > status_rank.get(agg["doc_status"], 0):
        agg["doc_status"] = rec["doc_status"]
        agg["account_prefix"] = rec["account_prefix"]
        agg["task_name"] = rec["task_name"]
    if not agg["code"] and rec["code"]:
        agg["code"] = rec["code"]
    # Merge doc_details (deduplicate by doc_id)
    existing_ids = {d["doc_id"] for d in agg["doc_details"]}
    for d in rec["doc_details"]:
        if d["doc_id"] not in existing_ids:
            agg["doc_details"].append(d)
            existing_ids.add(d["doc_id"])

# Convert to list, sort by minutes desc
items = []
for key, agg in aggregated.items():
    # Get representative activity name
    activity = agg["activities"][0]
    # Clean display name
    display = re.sub(r'\s+[A-Z]{6}\s*(\([^)]*\))?\s*$', '', activity).strip()
    if not display:
        display = activity
    items.append({
        "key": key,
        "display": display,
        "code": agg["code"],
        "minutes": agg["minutes"],
        "doc_status": agg["doc_status"],
        "account_prefix": agg["account_prefix"],
        "task_name": agg["task_name"],
        "doc_details": agg["doc_details"],
        "occurrences": len(agg["activities"]),
    })

items.sort(key=lambda x: x["minutes"], reverse=True)

# ── Statistics ─────────────────────────────────────────────────────────────────

total_mins = sum(r["minutes"] for r in records)
by_status = defaultdict(int)
for r in records:
    by_status[r["doc_status"]] += r["minutes"]

documented_mins = by_status["documented"]
mentioned_mins = by_status["mentioned"]
undocumented_mins = by_status["undocumented"]
unmatched_mins = by_status["unmatched"]

def pct(n):
    if total_mins == 0:
        return 0.0
    return round(100 * n / total_mins, 1)

# ── Oldest documented activities ───────────────────────────────────────────────

# Collect all doc_details with ages across items
all_doc_refs_used = []
for item in items:
    if item["doc_details"]:
        for dd in item["doc_details"]:
            if dd["age_years"] is not None:
                all_doc_refs_used.append({
                    "activity": item["display"],
                    "code": item["code"],
                    "doc_id": dd["doc_id"],
                    "title": dd["title"],
                    "last_modified": dd["last_modified"],
                    "age_years": dd["age_years"],
                    "match_type": dd["match_type"],
                    "activity_mins": item["minutes"],
                })
all_doc_refs_used.sort(key=lambda x: x["age_years"], reverse=True)
# Deduplicate by doc_id, keep highest age
seen_docs = {}
for d in all_doc_refs_used:
    if d["doc_id"] not in seen_docs:
        seen_docs[d["doc_id"]] = d
oldest_docs = sorted(seen_docs.values(), key=lambda x: x["age_years"], reverse=True)[:10]

# Biggest undocumented/unmatched time sinks
undoc_items = [i for i in items if i["doc_status"] in ("undocumented", "unmatched") and i["minutes"] > 0]
undoc_items.sort(key=lambda x: x["minutes"], reverse=True)

# ── Format report ──────────────────────────────────────────────────────────────

STATUS_ICON = {
    "documented": "✅",
    "mentioned": "⚠️ ",
    "undocumented": "❌",
    "unmatched": "🔍",
}
STATUS_LABEL = {
    "documented": "DOCUMENTED",
    "mentioned":  "MENTIONED ",
    "undocumented": "UNDOC     ",
    "unmatched":    "UNMATCHED ",
}

lines = []
lines.append("=" * 80)
lines.append("DAILY DOCUMENTATION COVERAGE REPORT — March 4, 2026")
lines.append("=" * 80)
lines.append(f"Generated: {TODAY.isoformat()}")
lines.append("")

lines.append("━" * 80)
lines.append("SUMMARY STATISTICS")
lines.append("━" * 80)
lines.append(f"  Total minutes logged        : {total_mins:>4} min")
lines.append(f"  ✅ Documented               : {documented_mins:>4} min  ({pct(documented_mins):>5.1f}%)")
lines.append(f"  ⚠️  Mentioned               : {mentioned_mins:>4} min  ({pct(mentioned_mins):>5.1f}%)")
lines.append(f"  ❌ Undocumented             : {undocumented_mins:>4} min  ({pct(undocumented_mins):>5.1f}%)")
lines.append(f"  🔍 Unmatched to task list   : {unmatched_mins:>4} min  ({pct(unmatched_mins):>5.1f}%)")
lines.append(f"  Coverage (doc+mentioned)    :       {pct(documented_mins + mentioned_mins):>5.1f}%")
lines.append("")

lines.append("━" * 80)
lines.append("FULL ACTIVITY LIST  (sorted by time spent, most first)")
lines.append("━" * 80)
lines.append(f"  {'ACTIVITY':<45} {'CODE':<6}  {'MIN':>4}  STATUS      DOCUMENTATION")
lines.append(f"  {'-'*44} {'-'*6}  {'-'*4}  {'-'*10}  {'-'*30}")

for item in items:
    icon = STATUS_ICON.get(item["doc_status"], "?")
    label = STATUS_LABEL.get(item["doc_status"], item["doc_status"])
    display = item["display"]
    if len(display) > 44:
        display = display[:43] + "…"
    code_str = item["code"] or "      "
    
    if item["doc_details"]:
        # Find the "best" doc (named match > account; lowest age is most relevant)
        named = [d for d in item["doc_details"] if d["match_type"] == "named"]
        best_docs = named if named else item["doc_details"]
        # Deduplicate by title
        seen_titles = {}
        for d in best_docs:
            if d["title"] not in seen_titles:
                seen_titles[d["title"]] = d
        best_docs = list(seen_titles.values())[:3]  # max 3
        
        first = True
        for d in best_docs:
            age_str = f"{d['age_years']:.1f}y" if d["age_years"] is not None else "?y"
            title_short = d["title"]
            if len(title_short) > 35:
                title_short = title_short[:34] + "…"
            if first:
                lines.append(f"  {display:<45} {code_str:<6}  {item['minutes']:>4}  {icon} {label}  [{d['doc_id']}] {title_short} ({age_str})")
                first = False
            else:
                lines.append(f"  {'':45} {'':6}  {'':4}  {'':13} [{d['doc_id']}] {title_short} ({age_str})")
    else:
        # No docs
        note = ""
        if item["doc_status"] == "unmatched":
            note = "→ not in task list"
            if item["code"]:
                note = f"→ code {item['code']} not in task list"
        elif item["doc_status"] == "undocumented":
            note = f"→ acct: {item['account_prefix']}" if item["account_prefix"] else "→ no account"
        lines.append(f"  {display:<45} {code_str:<6}  {item['minutes']:>4}  {icon} {label}  {note}")

lines.append("")

lines.append("━" * 80)
lines.append("BIGGEST UNDOCUMENTED / UNMATCHED TIME SINKS  (top 15)")
lines.append("━" * 80)
for i, item in enumerate(undoc_items[:15], 1):
    display = item["display"]
    if len(display) > 50:
        display = display[:49] + "…"
    code_str = f"[{item['code']}]" if item["code"] else "[------]"
    status = item["doc_status"].upper()
    acct = f"acct={item['account_prefix']}" if item["account_prefix"] else "no acct"
    lines.append(f"  {i:>2}. {display:<50} {code_str}  {item['minutes']:>4} min  {status}  {acct}")
lines.append("")

lines.append("━" * 80)
lines.append("OLDEST DOCUMENTATION REFERENCED TODAY  (top 10 by age)")
lines.append("━" * 80)
if oldest_docs:
    for d in oldest_docs:
        title_short = d["title"]
        if len(title_short) > 50:
            title_short = title_short[:49] + "…"
        match_lbl = f"[{d['match_type']}]"
        lines.append(f"  [{d['doc_id']}] {title_short}")
        lines.append(f"       Last modified: {d['last_modified']}  (age: {d['age_years']:.1f} years)  {match_lbl}")
        lines.append(f"       Referenced by: {d['activity']} ({d['activity_mins']} min)")
        lines.append("")
else:
    lines.append("  (No dated documentation found)")
lines.append("")

lines.append("━" * 80)
lines.append("NOTES")
lines.append("━" * 80)
lines.append("  Status definitions:")
lines.append("    ✅ DOCUMENTED  — task has a named or direct doc reference in master task list")
lines.append("    ⚠️  MENTIONED   — task is mentioned in docs for its account but not specifically described")
lines.append("    ❌ UNDOCUMENTED — task has no doc references at all")
lines.append("    🔍 UNMATCHED   — activity could not be matched to any task in master_task_list_v4")
lines.append("")
lines.append("  'Mentioned' docs are typically account-level documents that cover the area")
lines.append("  broadly (e.g. Dokumentation Lebenserhaltung) but don't describe this specific")
lines.append("  task. They count as weak coverage.")
lines.append("")

report_path = os.path.join(BASE, "data", "daily_doc_coverage_2026-03-04.txt")
with open(report_path, "w", encoding="utf-8") as f:
    f.write("\n".join(lines) + "\n")

print(f"Report written: {report_path}")
print()
print("=== QUICK SUMMARY ===")
print(f"Total minutes logged : {total_mins}")
print(f"✅ Documented        : {documented_mins} min ({pct(documented_mins):.1f}%)")
print(f"⚠️  Mentioned         : {mentioned_mins} min ({pct(mentioned_mins):.1f}%)")
print(f"❌ Undocumented      : {undocumented_mins} min ({pct(undocumented_mins):.1f}%)")
print(f"🔍 Unmatched         : {unmatched_mins} min ({pct(unmatched_mins):.1f}%)")
print(f"Coverage (doc+ment.) : {pct(documented_mins + mentioned_mins):.1f}%")
print()
print(f"Top 5 undocumented time sinks:")
for item in undoc_items[:5]:
    print(f"  {item['minutes']:>4} min  [{item['code'] or '------'}]  {item['display'][:55]}")
